api/src/Entity/ScheduleEntry.php

```diff
@@ -136,41 +130,54 @@ public function getCamp(): ?Camp {
     }
 
     /**
-     * The start date and time of the schedule entry. This is a read-only convenience property.
-     *
-     * @return null|DateTime
+     * Start date and time of the schedule entry.
      */
-    #[ApiProperty(writable: false, example: '2022-01-02T00:00:00+00:00', openapiContext: ['format' => 'date'])]
+    #[ApiProperty(example: '2022-01-02T00:00:00+00:00', required: true, openapiContext: ['format' => 'date-time'])]
+    #[Assert\GreaterThanOrEqual(propertyPath: 'period.start')]
     #[Groups(['read'])]
     public function getStart(): ?DateTime {
         try {
             $start = $this->period?->start ? DateTime::createFromInterface($this->period->start) : null;
-            $start?->add(new DateInterval('PT'.$this->periodOffset.'M'));
+            $start?->modify("{$this->startOffset} minutes");
 
             return $start;
         } catch (\Exception $e) {
             return null;
         }
     }
 
+    #[Groups(['write'])]
+    public function setStart(DateTimeInterface $start): void {
+        if (null !== $this->period?->start) {
```

Comment:
wenn man hier jetzt zu früh setStart aufruft, bekommen wir das nicht mit.
Wäre hier eine \RuntimeException angebracht?

Reply:
Dürfte eigentlich nicht passieren. Aber sicher safer, implemented in https://github.com/ecamp/ecamp3/pull/2484/commits/d1fff51c7f200db3ce7e1025ad89731e86cce9fb

Reply:
Ah, jetzt schlägt `CreateScheduleEntryTest::testCreateScheduleEntryValidatesMissingPeriod` fehlt und ich erinnere mich auch wieder, dass ich deshalb keine Exception geschmissen habe.

Problem ist halt, dass ich errorfrei durch das Deserialisieren kommen muss, damit ich überhaupt Validieren kann.

Ich frage mich, ob wir hier jetzt grad eine Abhängigkeit eingebaut haben, die wir eigentlich vermeiden wollten: period property muss deserialisiert werden, bevor start/end property deserialisiert werden können.

Reply:
Meines Wissens gibt es keine generische Möglichkeit, die Reihenfolge während dem Denormalizieren zu steuern. Die Reihenfolge scheint wird vom POST-Payload vorgegeben zu sein:

Kurz getestest + ist tatsächlich ein Problem: Je nach dem ob man `start` oder `period` zuerst im Payload erwähnt, gehen die Tests durch oder nicht.

Irgendwelche Vorschläge/Präferenzen:
- Manueller Denormalizer, welcher die Reihenfolge sicherstellt?
- Umstellen von offset auf Absolutwerte?
- Andere Optionen?

Reply:
Wenn wir wollen, dass das Frontend keine Offset-Brechnungen machen muss, sehe ich nur folgende beiden Varainten:
- Manueller Denormalizer (Vorschlag von @usu)
- Cache-Variable für ScheduleEntry Start/Ende  (siehe Skizze unten)

Ich sehe folgende Argumente
- (+) verständlicher: manueller Denormalizer ist zum lesen evtl. verständlicher
- (-) aufwändiger: manueller Denormalizer muss bei Schema-Anpassungen ebenfalls angepasst werden

```php

    // Stored in DB
    private Period $period;
    private int $startOffset;
    private int $endOffset;
    // Caches
    private DateTimeInterface $_start;
    private DateTimeInterface $_end;


    public function getPeriod(): Period {
        return $this->_period;
    }

    public function getStart(): DateTimeInterface  {
        // if value not yet cached, calc cached value now
        if (null == $this->_start && null != $this->period) {
            $this->_start = $this->period->start + $this->startOffset;
        }
        return $this->_start;
    }

    public function getEnd(): DateTimeInterface  {
        // if value not yet cached, calc cached value now
        if (null == $this->_end && null != $this->period) {
            $this->_end = $this->period->start + $this->endOffset;
        }
        return $this->_end;
    }

    public function setPeriod(Period $period): void {
        $this->period = $period;
        // if start has been denormalized, calc startOffset
        if (null != $this->_start) {
            $this->startOffset = $this->_start - $this->period->start;
        }
        // if end has been denormalized, calc endOffset
        if (null != $this->_end) {
            $this->endOffset = $this->end - $this->period->start;
        }
    }

    public function setStart(DateTimeInterface $start): void  {
        $this->_start = $start;
        if (null != $this->period) {
            $this->startOffset = $start - $this->period->start;
        }
    }

    public function setEnd(DateTimeInterface $end): void  {
        $this->_end = $end;
        if (null != $this->period) {
            $this->endOffset = $end - $this->period->start;
        }
    }

```

Reply:
Die weiteren Varianten wären noch:

- Einbauen, dass man abhängigkeiten beim deserialisieren angeben kann
- Die Reihenfolge beim deserialisieren angeben
- DTO verwenden

Ich würde vorschlagen, wir starten mal mit dem caching vorschlag. (Für das get müssen wir ja noch keinen cache einbauen.
Wenn wir nur logik in setStart, setEnd und setPeriod einbauen müssen, sollte es überschaubar bleiben.

Reply:
Danke für die Inputs. Vorschlag von Forte looks good to me. Implemented in https://github.com/ecamp/ecamp3/pull/2484/commits/8edffc58ecf0e73cb9d09e603a4d3a8541cc0e7a

Einziger Unterschied, dass ich in den gettern die internen Variablen nicht update. Getter mit Nebeneffekt bin ich eher skeptisch, und ist für die Funktionsweise auch nicht notwendig.